features/include/pcl/features/impl/gasd.hpp

```diff
@@ -0,0 +1,404 @@
+        hists[grid_idx + (grid_size + 3) * (grid_size + 2)][h_idx + 1] += v_xyzh1101;
+        hists[grid_idx + (grid_size + 3) * (grid_size + 2) + 1][h_idx] += v_xyzh1110;
+        hists[grid_idx + (grid_size + 3) * (grid_size + 2) + 1][h_idx + 1] += v_xyzh1111;
+
+        break;
+      }
+    }
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////
+  template <typename PointInT, typename PointOutT> void
+  GASDEstimation<PointInT, PointOutT>::copyHistogramsToOutput (int grid_size,
+                                                               int hists_size,
+                                                               vector<Eigen::VectorXf> &hists,
+                                                               PointCloudOut &output,
+                                                               int &pos)
+  {
+    for (int i = 0; i < grid_size; ++i)
+    {
+      for (int j = 0; j < grid_size; ++j)
+      {
+        for (int k = 0; k < grid_size; ++k)
+        {
+          int idx = ( (i + 1) * (grid_size + 2) + (j + 1)) * (grid_size + 2) + (k + 1);
+
+          for (int l = 1; l <= hists_size; ++l, ++pos)
+          {
+            output.points[0].histogram[pos] = hists[idx][l];
+          }
+        }
+      }
+    }
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////
+  template <typename PointInT, typename PointOutT> void
+  GASDEstimation<PointInT, PointOutT>::computeFeature (PointCloudOut &output)
+  {
+    // compute alignment transform using reference frame
+    computeAlignmentTransform ();
+
+    PointInTPtr shape_samples (new pcl::PointCloud<PointInT>);
+    pcl::PointCloud<pcl::PointXYZHSV>::Ptr color_samples;
+
+    // align point cloud
+    pcl::transformPointCloud (*surface_, *indices_, *shape_samples, transform_);
+
+    int shape_grid_size = shape_half_grid_size_ * 2;
+    int color_grid_size = color_half_grid_size_ * 2;
+
+    // each histogram dimension has 2 additional bins, 1 in each boundary, for performing interpolation
+    vector < Eigen::VectorXf > shape_hists ( (shape_grid_size + 2) * (shape_grid_size + 2) * (shape_grid_size + 2));
+    vector < Eigen::VectorXf > color_hists;
+
+    for (int i = 0; i < shape_hists.size (); ++i)
+    {
+      shape_hists[i].setZero (shape_hists_size_ + 2);
+    }
+
+    if (use_color_)
+    {
+      bool xyz_available = pcl::traits::has_xyz < PointInT > ::value;
+      bool rgb_available = pcl::traits::has_color < PointInT > ::value;
+
+      if (!xyz_available || !rgb_available)
+      {
+        PCL_ERROR ("[pcl::%s::computeFeature] No input dataset containing xyz and/or rgb was given!\n", getClassName ().c_str ());
+        return;
+      }
+
+      // RGB -> HSV conversion requires a PointXYZRGB point cloud
+      pcl::PointCloud<pcl::PointXYZRGB>::Ptr temp_samples (new pcl::PointCloud<pcl::PointXYZRGB>);
+
+      pcl::copyPointCloud (*shape_samples, *temp_samples);
```

Comment:
I'd say we better rewrite `PointCloudXYZRGBtoXYZHSV` to be templated on input point type and accept any type as long as it has `rgba` field. I can take care of this later this month.